addons/purchase_stock/models/purchase.py

```diff
@@ -355,8 +355,7 @@ def _prepare_stock_moves(self, picking):
 
         move_dests = self.move_dest_ids
         if not move_dests:
-            move_dests = self.move_ids.move_dest_ids
-
+            move_dests = self.move_ids.move_dest_ids.filtered(lambda m: not m.origin_returned_move_id)
```

Comment:
.filtered(lambda x: x.state != 'cancel' and not x.location_dest_id.usage == "supplier"):

+ add a test :)